### PR TITLE
Better std.algorithm.comparison.clamp error messages

### PR DESCRIPTION
**Pull request by burner**

### CODE REVIEW

File: changelog/clamp_assert_message.dd [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Better static assert messages for `std.algorithm.comparion.clamp`
```
Comment (RazvanN7) [inline changelog/clamp_assert_message.dd:1]:
```suggestion
Better static assert messages for `std.algorithm.comparison.clamp`
```

File: changelog/clamp_assert_message.dd [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+
+Up until now `clamp` used a template constraint to check if the passed types
```
Comment (RazvanN7) [inline changelog/clamp_assert_message.dd:3]:
```suggestion
Up until now, `clamp` used a template constraint to check if the passed types
```

File: changelog/clamp_assert_message.dd [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+could be used. If it were not, it was very tedious to figure out why.
```
Comment (RazvanN7) [inline changelog/clamp_assert_message.dd:4]:
```suggestion
could be used. If they were not, it was very tedious to figure out why.
```

File: changelog/clamp_assert_message.dd [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+As the template constraint is not used to overload the symbol template
```
Comment (RazvanN7) [inline changelog/clamp_assert_message.dd:6]:
```suggestion
As the template constraint is not used for overload resolution
```

File: changelog/clamp_assert_message.dd [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+function, the constrains are move into static asserts with expressive error
```
Comment (RazvanN7) [inline changelog/clamp_assert_message.dd:7]:
```suggestion
the constraints are moved into static asserts with expressive error
```

File: std/algorithm/comparison.d [hunk 6]
```diff
@@ -577,19 +577,26 @@ Returns:
...
+            , "' must be implicitly convertable to type of T1 '"
+            , T1.stringof, "'");
+    static assert(is(T3 : T1), "T3 of type '", T3.stringof
+            , "' must be implicitly convertable to type of T1 '"
+            , T1.stringof, "'");
+
+    // cannot use :
+    // `if (is(typeof(val.lessThan(lower) ? lower : val.greaterThan(upper) ? upper : val) : T1))
+    // because of https://issues.dlang.org/show_bug.cgi?id=16235.
+    // Once that is fixed, we can simply use the ternary in both the template constraint
```
Comment (RazvanN7) [inline std/algorithm/comparison.d:591]: This comment is not accurate anymore if the constraint is removed.

File: std/algorithm/comparison.d [hunk 7]
```diff
@@ -577,19 +577,26 @@ Returns:
...
+    // and the template body
+    static assert(is(typeof(val.lessThan(lower)
+        ? lower
+        : val.greaterThan(upper)
+            ? upper
+            : val)));
```
Comment (RazvanN7) [inline std/algorithm/comparison.d:597]: I am confused as to why this assert is used given that no error message is provided.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]

**Comment by burner** [comment]
@RazvanN7 awesome review thank you, arguable better than the PR itself.

File: std/algorithm/comparison.d [hunk 8]
```diff
@@ -577,19 +577,16 @@ Returns:
...
-    && (is(T2 : T1) && is(T3 : T1)))
-// cannot use :
-// `if (is(typeof(val.lessThan(lower) ? lower : val.greaterThan(upper) ? upper : val) : T1))
-// because of https://issues.dlang.org/show_bug.cgi?id=16235.
-// Once that is fixed, we can simply use the ternary in both the template constraint
-// and the template body
-in
 {
+    static assert(is(T2 : T1), "T2 of type '", T2.stringof
+            , "' must be implicitly convertable to type of T1 '"
```
Comment (RazvanN7) [inline std/algorithm/comparison.d:582]:
```suggestion
            , "' must be implicitly convertible to type of T1 '"
```

Comment (RazvanN7) [reply std/algorithm/comparison.d:582]: As per: https://whichiscorrect.com/convertible-vs-convertable/

**Review by RazvanN7 — COMMENTED** [review]
Thanks. A few nits and we are good to go.

**Review by RazvanN7 — APPROVED** [review]

**Comment by burner** [comment]
@RazvanN7 thank you

**Pull request merged**